File: src/levels/L19-event-bus/ts/facade.ts
// Cada servicio que produce algo llama a mano a cada servicio interesado.
class Orders {
  // region: Orders#place
  place(order: Event) {
    new Inventory().onEvent(order)
    new Billing().onEvent(order)
    new Loyalty().onEvent(order)
  }
  // endregion
}

class Returns {
  // Devoluciones copió la lista… y se le olvidó Facturación y Lealtad.
  // region: Returns#process
  process(returnRequest: Event) {
    new Inventory().onEvent(returnRequest)
  }
  // endregion
}

// Una fachada que agrupa los servicios… a la que cada productor sigue llamando con su propia lista.
// region: Services
class Services {
  notify(event: Event, ...targets: Subscriber[]) { targets.forEach((t) => t.onEvent(event)) }
}
// endregion

File: src/levels/L19-event-bus/facade.rb
# Cada servicio que produce algo llama a mano a cada servicio interesado.
class Orders
  # region: Orders#place
  def place(order)
    Inventory.new.on_event(order)
    Billing.new.on_event(order)
    Loyalty.new.on_event(order)
  end
  # endregion
end

class Returns
  # Devoluciones copió la lista… y se le olvidó Facturación y Lealtad.
  # region: Returns#process
  def process(return_request)
    Inventory.new.on_event(return_request)
  end
  # endregion
end

# Una fachada que agrupa los servicios… a la que cada productor sigue llamando con su propia lista.
# region: Services
class Services
  def notify(event, *targets) = targets.each { |t| t.on_event(event) }
end
# endregion
